### PR TITLE
Include folders in the result of getAdapterObjects

### PR DESCRIPTION
**Pull request by AlCalzone**
fixes: #1085 

I have a bunch of questions when looking at this code - lets discuss

### CODE REVIEW

File: lib/adapter.js [hunk 1]
```diff
@@ -1887,28 +1887,40 @@ function Adapter(options) {
...
+                        if (!adapterObjects) { // if objects is no longer existing, we do not need to unsubscribe
+                            return tools.maybeCallback(callback, objects);
                         }
-                        if (_devices && _devices.rows) {
-                            for (let d = _devices.rows.length - 1; d >= 0; d--) {
-                                objects[_devices.rows[d].id] = _devices.rows[d].value;
+
+                        adapterObjects.getObjectView('system', 'folder', {startkey: this.namespace + '.', endkey: this.namespace + '.\u9999', include_docs: true}, (err, _folders) => {
+                            if (_channels && _channels.rows) {
+                                for (let c = _channels.rows.length - 1; c >= 0; c--) {
```
Comment (AlCalzone) [inline lib/adapter.js:1897]: Why are these objects assigned in the reverse order?

File: lib/adapter.js [hunk 2]
```diff
@@ -1887,28 +1887,40 @@ function Adapter(options) {
...
+                                    objects[_channels.rows[c].id] = _channels.rows[c].value;
+                                }
                             }
-                        }
-                        if (_states && _states.rows) {
-                            if (options.states) {
-                                this.oStates = {};
+                            if (_devices && _devices.rows) {
+                                for (let d = _devices.rows.length - 1; d >= 0; d--) {
+                                    objects[_devices.rows[d].id] = _devices.rows[d].value;
```
Comment (AlCalzone) [inline lib/adapter.js:1903]: Why are channels added before devices?

File: lib/adapter.js [hunk 3]
```diff
@@ -1887,28 +1887,40 @@ function Adapter(options) {
...
+                            if (_folders && _folders.rows) {
+                                for (let f = _folders.rows.length - 1; f >= 0; f--) {
+                                    objects[_folders.rows[f].id] = _folders.rows[f].value;
                                 }
                             }
-                        }
-                        return tools.maybeCallback(callback, objects);
+                            if (_states && _states.rows) {
+                                if (options.states) {
+                                    this.oStates = {};
```
Comment (AlCalzone) [inline lib/adapter.js:1913]: What is this (and the lines below about)? Why are we nulling the states cache when reading all adapter objects?

File: lib/adapter.js [hunk 4]
```diff
@@ -1887,28 +1887,40 @@ function Adapter(options) {
                     }
 
                     adapterObjects.getObjectView('system', 'device', {startkey: this.namespace + '.', endkey: this.namespace + '.\u9999', include_docs: true}, (err, _devices) => {
-                        if (_channels && _channels.rows) {
-                            for (let c = _channels.rows.length - 1; c >= 0; c--) {
-                                objects[_channels.rows[c].id] = _channels.rows[c].value;
-                            }
+
```
Comment (AlCalzone) [inline lib/adapter.js:1890]: Should we bail early with an error when one of these calls returns an error?

**Comment by AlCalzone** [comment]
And most of all: why is this very specific behavior not commented? 🤐

File: lib/adapter.js (hunk 4 above)
Comment (AlCalzone) [reply lib/adapter.js:1890]: Hm that comment wasn't meant to go here... I mean the `err` in the getObjectView callback

**Comment by foxriver76** [comment]
Lgtm but would be great to implement new code in async style we have `objects.getObjectViewAsync` too.  Regarding your questions we most likely need info from  @GermanBluefox

**Comment by AlCalzone** [comment]
I've changed it as suggested and I've removed the back-to-front assignment since I don't see why it should be necessary.
Clearing oStates needs to be discussed IMO

**Comment by AlCalzone** [comment]
And the test failures **seem** unrelated but I'm not sure.

**Comment by foxriver76** [comment]
First test failing at https://github.com/ioBroker/ioBroker.js-controller/blob/36549540d4a51bc7212bc2f5188de12a51a6d264/test/lib/testObjectsFunctions.js#L111 is susp

**Comment by AlCalzone** [comment]
I'll check it later, thanks

**Comment by AlCalzone** [comment]
Mehhh:
The error is: `Cannot find search "folder" in "system"`. Is that something bugged in the tests or do we need to add a method for that to the DB?

**Comment by foxriver76** [comment]
I've made a PR against your branch to fix it. It's only a testing issue, the view is already existing in production io-pack https://github.com/ioBroker/ioBroker.js-controller/blob/36549540d4a51bc7212bc2f5188de12a51a6d264/io-package.json#L276

**Comment by GermanBluefox** [comment]
Thank you!

**Pull request merged**